docs/storage.md

```diff
@@ -0,0 +1,187 @@
+- Segments with only the magic header (no data) are deleted to avoid uploading empty files.
+
+---
+
+## WAL Segment Rotation and Backpressure: Flow Diagram
+
+```mermaid
+flowchart TD
+    A[Receive Telemetry] --> B[Append to WAL Segment]
+    B --> C{Segment Max Size or Age Reached?}
+    C -- No --> B
+    C -- Yes --> D[Rotate Segment]
+    D --> E{Segment < Transfer Threshold?}
+    E -- Yes --> F[Transfer to Peer]
+    E -- No --> G[Upload to ADX]
+    F --> H[Delete Segment]
+    G --> H
+    B --> I{Disk Usage/Segment Count Exceeded?}
+    I -- No --> B
+    I -- Yes --> J[Reject Write/Transfer]
+```
+
+---
+
+## Key Code Mechanisms
+
+- **Rotation**: `rotateSegmentIfNecessary()` in `pkg/wal/wal.go` handles segment rotation and cleanup.
+- **Backpressure**: `validateLimits()` in `pkg/wal/wal.go` checks disk usage and segment count before writes.
+- **Segment Transfer/Upload**: Logic in `ingestor/cluster/batcher.go` and `ingestor/adx/uploader.go` manages peer transfer and ADX upload.
+- **Cleanup**: Segments are deleted after upload/transfer; empty segments are removed on startup.
+
+---
+
+## Configuration
+
+- Limits and thresholds are set via CLI flags and/or Ingestor CRD fields:
+  - `--storage-dir`: Directory for WAL files (required)
+  - `--max-disk-usage`: Max bytes for WAL storage (**default: 10 GiB if unset**)
+  - `--max-segment-count`: Max number of WAL segments (**default: 10,000 if unset**)
+  - `--max-segment-size`: Max size per segment
+  - `--max-segment-age`: Max age per segment
+  - `--max-transfer-age`: Max age for peer transfer
+  - `--enable-wal-fsync`: Enable fsync for durability
+
+> **Note:**
+> - **Ingestor** defaults to a maximum disk usage of **10 GiB** (`--max-disk-usage=10*1024*1024*1024`) if not set.
+> - **Collector** defaults to **unlimited** disk usage (`max-disk-usage=0`) if not set. You must explicitly set a limit to prevent disk exhaustion.
+
+---
+
+## References
+- [WAL Segment File Format](concepts.md#wal-segment-file-format)
+- [Ingestor Data Flow](ingestor.md#data-flow)
+- [Azure Data Explorer Ingestion Best Practices](https://docs.microsoft.com/en-us/azure/data-explorer/ingest-best-practices)
+
+---
+
+## Summary Table
+
+| Mechanism         | Purpose                                 | Triggered By                |
+|-------------------|-----------------------------------------|-----------------------------|
+| WAL Write         | Durability, batching                    | Every telemetry write       |
+| Segment Rotation  | Batching, upload/transfer optimization  | Max size/age, periodic task |
+| Peer Transfer     | Coalescing small files                  | Segment < threshold         |
+| Direct Upload     | Timely upload, fallback                 | Segment large/old/transfer fail |
+| Cleanup           | Free disk, prevent storage exhaustion   | After upload/transfer       |
+| Backpressure      | Prevent disk exhaustion, signal overload| On write/transfer if limits exceeded |
```

Comment:
Not quite sure about this table - seems to be combining several unrelated things?